resource/resource.go

```diff
@@ -0,0 +1,166 @@
+package resource
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+const (
+	envVarType = "OC_RESOURCE_TYPE"
+	envVarTags = "OC_RESOURCE_TAGS"
+)
+
+// Resource describes an entity about which data is exposed.
+type Resource struct {
+	Type string
+	Tags map[string]string
+}
+
+// EncodeTags encodes a tags to a string as provided via the OC_RESOURCE_TAGS environment variable.
+func EncodeTags(tags map[string]string) string {
+	sortedKeys := make([]string, 0, len(tags))
+	for k := range tags {
+		sortedKeys = append(sortedKeys, k)
+	}
+	sort.Strings(sortedKeys)
+
+	s := ""
+	for i, k := range sortedKeys {
+		if i > 0 {
+			s += ","
+		}
+		s += k + "=" + strconv.Quote(tags[k])
+	}
+	return s
+}
+
+var tagRegex = regexp.MustCompile(`\s*([a-zA-Z0-9-_./]+)=(?:(".*?")|([^\s="]+))\s*,`)
+
+// DecodeTags decodes serialized a tag map as used in the OC_RESOURCE_TAGS variable.
+// A list of tags of the form `<key1>=<value1>,<key2>=<value2>,...` is accepted.
+// Domain names and paths are accepted as tag keys. Values may be quoted or unquoted in general.
+// If a value contains whitespaces, =, or " characters, it must always be quoted.
+func DecodeTags(s string) (map[string]string, error) {
+	m := map[string]string{}
+	// Ensure a trailing comma, which allows us to keep the regex simpler
+	s = strings.TrimRight(strings.TrimSpace(s), ",") + ","
+
+	for len(s) > 0 {
+		match := tagRegex.FindStringSubmatch(s)
+		if len(match) == 0 {
+			return nil, fmt.Errorf("invalid tag formatting, remainder: %s", s)
+		}
+		v := match[2]
+		if v == "" {
+			v = match[3]
+		} else {
+			var err error
+			if v, err = strconv.Unquote(v); err != nil {
+				return nil, fmt.Errorf("invalid tag formatting, remainder: %s, err: %s", s, err)
+			}
+		}
+		m[match[1]] = v
+
+		s = s[len(match[0]):]
+	}
+	return m, nil
+}
+
+// FromEnv loads resource information from the OC_TYPE and OC_RESOURCE_TAGS environment variables.
+func FromEnv(context.Context) (*Resource, error) {
+	res := &Resource{
+		Type: strings.TrimSpace(os.Getenv(envVarType)),
+	}
+	tags := strings.TrimSpace(os.Getenv(envVarTags))
+	if tags == "" {
+		return res, nil
+	}
+	var err error
+	if res.Tags, err = DecodeTags(tags); err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// Merge resource information from b into a. In case of a collision, a takes precedence.
+func Merge(a, b *Resource) *Resource {
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	res := &Resource{
+		Type: a.Type,
+		Tags: map[string]string{},
+	}
+	for k, v := range a.Tags {
+		res.Tags[k] = v
+	}
+	if res.Type == "" {
+		res.Type = b.Type
+	}
+	for k, v := range b.Tags {
+		if _, ok := res.Tags[k]; !ok {
+			res.Tags[k] = v
+		}
+	}
+	return res
+}
+
+// Detector attempts to detect resource information.
+// If the detector cannot find specific information, the respective Resource fields should
+// be left empty but no error should be returned.
```

Comment:
This seems to imply its not valid to return `nil, nil` (saying that you should leave fields empty), though that seems like it would be the most natural way to return "this detector does not apply in the current environment".